fluent-bit/fluent-bit-go-loki/out_loki.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import "github.com/fluent/fluent-bit-go/output"
+import "github.com/grafana/loki/pkg/promtail/client"
+import "github.com/sirupsen/logrus"
+import kit "github.com/go-kit/kit/log/logrus"
+import "github.com/prometheus/common/model"
+import "github.com/cortexproject/cortex/pkg/util/flagext"
+import jsoniter "github.com/json-iterator/go"
+
+import (
+	"C"
+	"fmt"
+	"os"
+	"time"
+	"unsafe"
+)
+
+var loki client.Client
+var ls model.LabelSet
+var plugin GoOutputPlugin = &fluentPlugin{}
+
+type GoOutputPlugin interface {
+	PluginConfigKey(ctx unsafe.Pointer, key string) string
+	Unregister(ctx unsafe.Pointer)
+	GetRecord(dec *output.FLBDecoder) (ret int, ts interface{}, rec map[interface{}]interface{})
+	NewDecoder(data unsafe.Pointer, length int) *output.FLBDecoder
+	HandleLine(ls model.LabelSet, timestamp time.Time, line string) error
+	Exit(code int)
+}
+
+type fluentPlugin struct{}
+
+func (p *fluentPlugin) PluginConfigKey(ctx unsafe.Pointer, key string) string {
+	return output.FLBPluginConfigKey(ctx, key)
+}
+
+func (p *fluentPlugin) Unregister(ctx unsafe.Pointer) {
+	output.FLBPluginUnregister(ctx)
+}
+
+func (p *fluentPlugin) GetRecord(dec *output.FLBDecoder) (int, interface{}, map[interface{}]interface{}) {
+	return output.GetRecord(dec)
+}
+
+func (p *fluentPlugin) NewDecoder(data unsafe.Pointer, length int) *output.FLBDecoder {
+	return output.NewDecoder(data, length)
+}
+
+func (p *fluentPlugin) Exit(code int) {
+	os.Exit(code)
+}
+
+func (p *fluentPlugin) HandleLine(ls model.LabelSet, timestamp time.Time, line string) error {
+	return loki.Handle(ls, timestamp, line)
+}
+
+//export FLBPluginRegister
+func FLBPluginRegister(ctx unsafe.Pointer) int {
+	return output.FLBPluginRegister(ctx, "loki", "Loki Output plugin written in GO!")
```

Comment:
Maybe this should be something more descriptive, e.g. `Output plugin for Grafana Loki` or `Ship logs to Grafana Loki`